compiler/src/iree/compiler/Dialect/HAL/Conversion/StreamToHAL/ConvertStreamToHAL.cpp

```diff
@@ -44,6 +45,62 @@ static Value getOrCreateWaitFence(Location loc, Value timepointFence,
       loc, builder.getType<IREE::HAL::FenceType>());
 }
 
+// Finds a !hal.fence bound to |timepoint| via a chain op and returns it if
+// it is usable at the builder insertion point. The chain ops is only used if
+// it is the only consumer of the timepoint and it is removed upon return.
+static Value consumeBoundFence(Value timepoint,
+                               ConversionPatternRewriter &rewriter) {
+  // Must only have one use. We can't consume a fence multiple times.
+  if (!timepoint.hasOneUse()) return nullptr;  // >1 use
+
+  // The use must be an export to a fence.
+  auto chainOp = dyn_cast<IREE::Stream::TimepointChainExternalOp>(
+      *timepoint.getUsers().begin());
+  if (!chainOp) return nullptr;  // non-export use
+  assert(!chainOp.getExternalValues().empty());
+  auto fence = chainOp.getExternalValues().front();
+  if (!fence || !fence.getType().isa<IREE::HAL::FenceType>()) return nullptr;
+
+  // Try really hard to figure out if the fence can be used. A larger analysis
+  // pass running prior to conversion that did some code motion could help
+  // ensure the fence SSA value is usable in the places it is needed - for now
+  // we just do this local check that satisfies most common programs today. IPO
+  // would do something like add the fence as an argument to function calls so
+  // that the functions could consume it but inlining is pretty aggressive now.
+  if (!IREE::Util::isValueUsableForOp(fence, rewriter.getBlock(),
```

Comment:
Am I correct in assuming that this precisely covers the usage added in the wrap pass? (Ie. Doing more complicated/ad-hoc use by the user would necessitate precise analysis?

Reply:
Mostly - the nice thing about this path is that it's just an optimization: if we can't find a usable fence then we insert a host/device sync before signaling the fence the user provided (enough to meet the API contract). The thing that a more general analysis and sinking of the user-provided fence into call trees would provide is a greater chance of wait-free behavior especially in complex programs with flow control and async global variables. Will be cool, but thankfully not needed to start :)